rpm/0005-Retain-symlink-chains.patch

```diff
@@ -0,0 +1,209 @@
+-		free(use_target);
+-		use_target = tmp;
+-	}
+-	else
+-		errExit("asprintf");
+-
+-done:
+-	if (proc_pid)
+-		free(proc_pid);
+-	return use_target;
+-}
+-
+-void copy_link(const char *target, const char *linkpath, mode_t mode, uid_t uid, gid_t gid) {
+-	(void) mode;
+-	(void) uid;
+-	(void) gid;
++void copy_link(const char *target, const char *linkpath) {
++	int failed = 1;
++	char *linkdata = NULL;
+ 
+ 	// if the link is already there, don't create it
+ 	struct stat s;
+ 	if (lstat(linkpath, &s) == 0)
+-	       return;
+-
+-	char *rp = proc_pid_to_self(target);
+-	if (rp) {
+-		if (symlink(rp, linkpath) == -1) {
+-			free(rp);
+-			goto errout;
+-		}
+-		free(rp);
++		goto success;
++
++	// read source symlink
++	if (lstat(target, &s) == -1)
++		goto failure;
++
++	if (!S_ISLNK(s.st_mode))
++		goto failure;
++
++	ssize_t linksize = s.st_size ? (s.st_size + 1) : PATH_MAX;
++	if (!(linkdata = malloc(linksize)))
++	    goto failure;
++
++	ssize_t rc = readlink(target, linkdata, linksize);
++	if (rc < 0 ) {
++		if (!arg_quiet)
++			fprintf(stderr, "Error fcopy: readlink %s failed: %m\n", target);
++		goto failure;
++	}
++
++	if (rc >= linksize) {
++		if (!arg_quiet)
++			fprintf(stderr, "Error fcopy: readlink %s buffer overflow\n", target);
++		goto failure;
+ 	}
+-	else
+-		goto errout;
+ 
+-	return;
+-errout:
+-	if (!arg_quiet)
+-		fprintf(stderr, "Warning fcopy: cannot create symbolic link %s\n", target);
++	linkdata[rc] = 0;
++
++	// duplicate at the given path
++	if (symlink(linkdata, linkpath) == -1) {
++		if (!arg_quiet)
++			fprintf(stderr, "Error fcopy: creating %s symlink failed: %m\n", linkpath);
++		goto failure;
++	}
++
++	if (arg_debug)
++		fprintf(stderr, "fcopy: created symlink: %s -> %s\n", linkpath, linkdata);
++
++success:
++	failed = 0;
++failure:
++	if (failed && !arg_quiet)
++		fprintf(stderr, "Warning fcopy: cannot create symbolic link %s\n", linkpath);
++
++	free(linkdata);
+ }
+ 
+ 
+@@ -310,7 +292,7 @@ static int fs_copydir(const char *infname, const struct stat *st, int ftype, str
+ 		mkdir_attr(outfname, mode, uid, gid);
+ 	}
+ 	else if (ftype == FTW_SL) {
+-		copy_link(infname, outfname, mode, uid, gid);
++		copy_link(infname, outfname);
+ 	}
+ out:
+ 	free(outfname);
+@@ -396,22 +378,16 @@ static void duplicate_file(const char *src, const char *dest, struct stat *s) {
+ static void duplicate_link(const char *src, const char *dest, struct stat *s) {
+ 	char *rsrc = check(src);		  // we drop the result and use the original name
+ 	char *rdest = check(dest);
+-	uid_t uid = s->st_uid;
+-	gid_t gid = s->st_gid;
+-	mode_t mode = s->st_mode;
+ 
+-	// build destination file name
+-	char *name;
+-	//     char *ptr = strrchr(rsrc, '/');
+-	char *ptr = strrchr(src, '/');
+-	ptr++;
+-	if (asprintf(&name, "%s/%s", rdest, ptr) == -1)
+-		errExit("asprintf");
++	const char *linkname = strrchr(src, '/');
++	if (linkname++) {
++		char *linkpath = NULL;
++		if (asprintf(&linkpath, "%s/%s", rdest, linkname) == -1)
```

Comment:
Wouldn't this now result in a string with two '/' characters in a row, e.g. /path/to/smth//link?